LoRaEngine/test/LoRaWanNetworkServer.Test/DeduplicationStrategy_End2End_Tests.cs

```diff
@@ -0,0 +1,142 @@
+        [InlineData(DeduplicationMode.None, true)]
+        [InlineData(DeduplicationMode.Mark, false)]
+        [InlineData(DeduplicationMode.Drop, false)]
+        [InlineData(DeduplicationMode.None, false)]
+        public async Task Validate_Dup_Message_Processing(DeduplicationMode mode, bool confirmed)
+        {
+            var simulatedDevice = new SimulatedDevice(TestDeviceInfo.CreateABPDevice(1));
+            bool messageProcessed = mode == DeduplicationMode.Drop;
+            messageProcessed = false;
+            this.LoRaDeviceApi
+                .Setup(x => x.CheckDuplicateMsgAsync(It.IsAny<string>(), It.IsAny<int>(), It.IsAny<string>(), It.IsAny<int?>()))
+                .Returns<string, int, string, int?>((dev, fcntup, gateway, fcntdown) =>
+                {
+                    var isDup = messageProcessed;
+                    messageProcessed = true;
+                    Assert.True((confirmed && fcntdown.HasValue) || (!confirmed && !fcntdown.HasValue));
+                    return Task.FromResult<DeduplicationResult>(new DeduplicationResult
+                    {
+                        IsDuplicate = isDup,
+                        ClientFCntDown = fcntdown
+                    });
+                });
+
+            this.LoRaDeviceApi
+                .Setup(x => x.NextFCntDownAsync(It.IsAny<string>(), It.IsAny<int>(), It.IsAny<int>(), It.IsAny<string>()))
+                .ReturnsAsync((ushort)(simulatedDevice.FrmCntDown + 1))
+                .Callback(() =>
+                {
+                    // this call should only be made, if we do not have a deduplication strategy
+                    // since otherwise we expect the fcntDown to be calculated in the same API call as deduplication
+                    Assert.True(mode == DeduplicationMode.None);
+                });
+
+            var shouldBeMarked = false;
+
+            this.LoRaDeviceClient
+                .Setup(x => x.SendEventAsync(It.IsNotNull<LoRaDeviceTelemetry>(), null))
+                .ReturnsAsync(true)
+                .Callback<LoRaDeviceTelemetry, Dictionary<string, string>>((telemetry, dict) =>
+                {
+                    if (shouldBeMarked)
+                    {
+                        Assert.True(telemetry.DupMsg);
+                    }
+                    else
+                    {
+                        Assert.Null(telemetry.DupMsg);
+                    }
+                    shouldBeMarked = mode == DeduplicationMode.Mark;
+                });
+
+            this.LoRaDeviceClient
+                .Setup(x => x.ReceiveAsync(It.IsAny<TimeSpan>()))
+                .ReturnsAsync((Message)null);
+
+            await this.SendTwoMessages(mode, confirmed);
+        }
+
+        private async Task SendTwoMessages(DeduplicationMode mode, bool confirmed)
+        {
+            var simulatedDevice = new SimulatedDevice(TestDeviceInfo.CreateABPDevice(1));
+
+            var loRaDevice = this.CreateLoRaDevice(simulatedDevice);
```

Comment:
If you are trying to simulate parallel processing from multiple gateways you should create 2 LoRaDevice instances. Reusing the same will cause the requests to be queued under the unique instance.

Reply:
I know what you mean. If you really want it to be completely separate I would also have to duplicate the API services. However, I don't think that makes a difference in the Unit test as it all comes down to how the strategy handle the response from the API service and how the processing proceeds. If the API service mock handles that, I'm not sure having multiple devices will give us anything else in the test?

Reply:
Makes sense. I was not sure if you wanted to test processing the message in parallel vs handling the responses correctly. But, that would be more something to be done at the function level, but unfortunately we don't have any, right?